File: sdks/js/src/show.ts
/**
 * Show Control convenience methods for Maestra SDK.
 * Manages show lifecycle phases: idle, pre_show, active, paused, post_show, shutdown
 */

export type ShowPhase = 'idle' | 'pre_show' | 'active' | 'paused' | 'post_show' | 'shutdown'

export interface ShowState {
  phase: ShowPhase
  previous_phase: ShowPhase | null
  transition_time: string
  source: string | null
  context: Record<string, unknown>
}

export interface ShowTransitionResponse {
  phase: ShowPhase
  previous_phase: ShowPhase | null
  transition_time: string
  source: string | null
  context: Record<string, unknown>
}

export interface ShowHistoryEntry {
  from_phase: ShowPhase
  to_phase: ShowPhase
  source: string | null
  transition_time: string
  context: Record<string, unknown>
}

export type ShowStateCallback = (state: ShowState) => void

/**
 * Show Control
 * Provides convenience methods for managing show lifecycle phases.
 *
 * Usage:
 *   import { ShowControl } from '@maestra/sdk'
 *
 *   const show = new ShowControl('http://localhost:8080')
 *
 *   // Check current state
 *   const state = await show.getState()
 *   console.log(`Phase: ${state.phase}`)
 *
 *   // Run through show lifecycle
 *   await show.warmup()   // idle -> pre_show
 *   await show.go()       // pre_show -> active
 *   await show.pause()    // active -> paused
 *   await show.resume()   // paused -> active
 *   await show.stop()     // active -> post_show
 *   await show.reset()    // any -> idle
 *
 *   // Or transition directly
 *   await show.transition('active', 'stage-manager')
 *
 *   // View history
 *   const history = await show.getHistory(10)
 *
 *   // Subscribe to state changes (polling)
 *   const unsub = show.onChange((state) => {
 *     console.log(`Phase changed: ${state.phase}`)
 *   })
 *   // ... later ...
 *   unsub()
 */
export class ShowControl {
  private baseUrl: string
  private _callbacks: ShowStateCallback[] = []
  private _pollInterval: ReturnType<typeof setInterval> | null = null
  private _lastPhase: ShowPhase | null = null

  constructor(baseUrl: string) {
    this.baseUrl = baseUrl.replace(/\/$/, '')
  }

  private async request<T>(method: string, endpoint: string, body?: unknown): Promise<T> {
    const url = `${this.baseUrl}${endpoint}`
    const options: RequestInit = {
      method,
      headers: { 'Content-Type': 'application/json' },
    }
    if (body) {
      options.body = JSON.stringify(body)
    }

    const response = await fetch(url, options)
    if (!response.ok) {
      const error = await response.json().catch(() => ({ detail: 'Unknown error' }))
      throw new Error(error.detail || `HTTP ${response.status}`)
    }
    return response.json()
  }

  /** Get current show state */
  async getState(): Promise<ShowState> {
    return this.request('GET', '/show/state')
  }

  /** Transition to pre_show phase (warmup) */
  async warmup(): Promise<ShowTransitionResponse> {
    return this.request('POST', '/show/warmup')
  }

  /** Transition to active phase (go live) */
  async go(): Promise<ShowTransitionResponse> {
    return this.request('POST', '/show/go')
  }

  /** Pause the active show */
  async pause(): Promise<ShowTransitionResponse> {
    return this.request('POST', '/show/pause')
  }

  /** Resume a paused show */
  async resume(): Promise<ShowTransitionResponse> {
    return this.request('POST', '/show/resume')
  }

  /** Stop the show (transition to post_show) */
  async stop(): Promise<ShowTransitionResponse> {
    return this.request('POST', '/show/stop')
  }

  /** Shutdown the show (transition to shutdown phase) */
  async shutdown(): Promise<ShowTransitionResponse> {
    return this.request('POST', '/show/shutdown')
  }

  /** Reset the show back to idle */
  async reset(): Promise<ShowTransitionResponse> {
    return this.request('POST', '/show/reset')
  }

  /** Transition to an arbitrary phase */
  async transition(to: string, source = 'js-sdk'): Promise<ShowTransitionResponse> {
    return this.request('POST', '/show/transition', { to, source })
  }

  /** Get show transition history */
  async getHistory(limit = 20): Promise<ShowHistoryEntry[]> {
    return this.request('GET', `/show/history?limit=${limit}`)
  }

  /**
   * Subscribe to show state changes via polling.
   * The callback fires whenever the phase changes.
   *
   * @param callback - Called with the new ShowState when the phase changes
   * @param intervalMs - Polling interval in milliseconds (default: 1000)
   * @returns Unsubscribe function that stops polling when no callbacks remain
   */
  onChange(callback: ShowStateCallback, intervalMs = 1000): () => void {
    this._callbacks.push(callback)

    // Start polling if not already running
    if (!this._pollInterval) {
      this._pollInterval = setInterval(async () => {
        try {
          const state = await this.getState()
          if (this._lastPhase !== null && state.phase !== this._lastPhase) {
            this._callbacks.forEach((cb) => {
              try {
                cb(state)
              } catch (e) {
                console.error('Error in show state change callback:', e)
              }
            })
          }
          this._lastPhase = state.phase
        } catch (e) {
          // Silently ignore poll errors to avoid flooding the console
        }
      }, intervalMs)
    }

    // Return unsubscribe function
    return () => {
      const index = this._callbacks.indexOf(callback)
      if (index > -1) {
        this._callbacks.splice(index, 1)
      }
      // Stop polling when no callbacks remain
      if (this._callbacks.length === 0 && this._pollInterval) {
        clearInterval(this._pollInterval)
        this._pollInterval = null
        this._lastPhase = null
      }
    }
  }

  /** Stop all polling and clear callbacks */
  dispose(): void {
    if (this._pollInterval) {
      clearInterval(this._pollInterval)
      this._pollInterval = null
    }
    this._callbacks = []
    this._lastPhase = null
  }
}

File: sdks/python/maestra/show.py
"""
Show Control convenience methods for Maestra SDK.
Provides ShowControl class for managing show lifecycle phases.

Phases: idle, pre_show, active, paused, post_show, shutdown
"""

from typing import Dict, Any, Optional, List, TYPE_CHECKING

if TYPE_CHECKING:
    from .client import HttpTransport


class ShowControl:
    """
    Show control convenience class for managing show lifecycle.

    Usage:
        from maestra import MaestraClient

        client = MaestraClient()
        await client.connect()

        show = ShowControl(client._http)

        # Check current state
        state = await show.get_state()
        print(f"Phase: {state['phase']}")

        # Run through show lifecycle
        await show.warmup()   # idle -> pre_show
        await show.go()       # pre_show -> active
        await show.pause()    # active -> paused
        await show.resume()   # paused -> active
        await show.stop()     # active -> post_show
        await show.reset()    # any -> idle

        # Or transition directly
        await show.transition("active", source="stage-manager")

        # View history
        history = await show.get_history(limit=10)
    """

    def __init__(self, transport: "HttpTransport"):
        self._transport = transport

    async def get_state(self) -> Dict[str, Any]:
        """Get current show state.

        Returns:
            dict with keys: phase, previous_phase, transition_time, source, context
        """
        return await self._transport._request("GET", "/show/state")

    async def warmup(self) -> Dict[str, Any]:
        """Transition to pre_show phase (warmup).

        Returns:
            dict with updated show state
        """
        return await self._transport._request("POST", "/show/warmup")

    async def go(self) -> Dict[str, Any]:
        """Transition to active phase (go live).

        Returns:
            dict with updated show state
        """
        return await self._transport._request("POST", "/show/go")

    async def pause(self) -> Dict[str, Any]:
        """Pause the active show.

        Returns:
            dict with updated show state
        """
        return await self._transport._request("POST", "/show/pause")

    async def resume(self) -> Dict[str, Any]:
        """Resume a paused show.

        Returns:
            dict with updated show state
        """
        return await self._transport._request("POST", "/show/resume")

    async def stop(self) -> Dict[str, Any]:
        """Stop the show (transition to post_show).

        Returns:
            dict with updated show state
        """
        return await self._transport._request("POST", "/show/stop")

    async def shutdown(self) -> Dict[str, Any]:
        """Shutdown the show (transition to shutdown phase).

        Returns:
            dict with updated show state
        """
        return await self._transport._request("POST", "/show/shutdown")

    async def reset(self) -> Dict[str, Any]:
        """Reset the show back to idle.

        Returns:
            dict with updated show state
        """
        return await self._transport._request("POST", "/show/reset")

    async def transition(self, to: str, source: str = "python-sdk") -> Dict[str, Any]:
        """Transition to an arbitrary phase.

        Args:
            to: Target phase (idle, pre_show, active, paused, post_show, shutdown)
            source: Identifier for the source of this transition

        Returns:
            dict with updated show state
        """
        return await self._transport._request(
            "POST", "/show/transition", json={"to": to, "source": source}
        )

    async def get_history(self, limit: int = 20) -> List[Dict[str, Any]]:
        """Get show transition history.

        Args:
            limit: Maximum number of history entries to return

        Returns:
            List of transition history entries
        """
        return await self._transport._request(
            "GET", "/show/history", params={"limit": limit}
        )
